### ISSUE TITLE
hierarchy isn't saved 

### ISSUE THREAD

**Issue by pjarnfelt** [radiatoryang/merino#3]
In the Node list (which you can reorder), you can add a node underneath another node. This is purely for organizational purposes, but quite useful. This structure does not get saved into the file. 
![structure](https://user-images.githubusercontent.com/891674/44781305-4b2be200-ab84-11e8-8400-e8215c33fdc0.PNG)

**Comment by radiatoryang** [comment]
so this one's a bit tricky -- there's no structure in the Yarn file spec that actually supports this, and I wouldn't want to create Yarn files that aren't to spec

the one thing I could imagine doing: add a "ParentNode:[nodename]" to the node's tag list? that way, every other Yarn editor or parser can ignore it as a tag

**Comment by SimianLogic** [comment]
This metadata seems like it belongs to Merino instead of Yarn.

First thought: you currently have one leaf type (TreeElement): Yarn Node. A single yarn file gets expanded out to fill the tree with no structure retained.  You could instead have leaves for Folder, YarnFile, and YarnNode:
- Pressing "save" would save just the YarnFile for the active node. 
- Dragging a YarnNode from one YarnFile to another within tree view would transfer the node to the drop target and save both files. 
- Dragging a YarnNode into a Folder within tree view would create a new YarnFile with just that YarnNode (probably with the name of the node as a default filename). 
- Your "open" asset chooser could load either a single yarn file (same as now) or a folder (and then just recursively scan for yarn files to build the tree). 

Second thought: write a ScriptableObject (or even just JSON file) that gets saved alongside yarn files and adds Merino-specific metadata. This seems easier but somehow grosser? Large projects might run into scale problems with a monolithic yarn file and enforcing/maintaining the metadata structure seems harder than just reading what's on the file system.

**Comment by pjarnfelt** [comment]
Making Folders and files for each node seems super messy to me. Making ScriptableObjects or JSON files next to the yarn files seems ugly to me also. The beauty of Merino is that I can contain everything inside one yarn file that we can even open in a text editor. It makes it better/easier for my non technically skilled (or non-organized) text writers to handle. 
The metadata ParentNode would totally accomplish it, and yea, it augments the yarn structure, but isn't that what metadata is for? <blink> <blink>

**Comment by SimianLogic** [comment]
You wouldn't make a new file or folder for every node, just as-needed for organization. If I was building a little serial with two character choices and three scenes my folder structure might look like:

- root (folder)
- - bob (folder)
- - - scene_1.yarn (with as many nodes as are in the scene)
- - - scene_2.yarn
- - - scene_3.yarn
- - sally (folder)
- - - scene_1.yarn
- - - scene_2.yarn
- - - scene_3.yarn

This is the actual filesystem -- in merino, you'd see a leaf for each node in the yarn file. Possible workflows would include:

- work on a single scene (just open a yarn file)
- work on a whole character's sequence (open that character's folder, move nodes around between scenes. more "edit" mode than "author")
- I'm not sure I ever need to open the root folder in this example, but in a more complicated example where I'm doing real branching I might want to split each story branch into a folder

Nothing in this workflow prevents me from opening yarn files in a text editor (or opening the whole root folder in something like SublimeText or Textmate) if I want a file tree or tabbed editing.

After writing all this out I'd say the biggest argument against it is: *do I actually need to have more than 1 yarn file open at a time?* You would gain the ability to preserve state across file boundaries (i.e. read a variable set in one yarn file after you jump to a node in a new yarn file), get some editing convenience (move nodes between files, move nodes into new files), and allow people to work simultaneously without getting into merge hell. 

The first two are fixing problems I created (if it's all one monolithic Yarn file, both of these are free). The final one is a nice-to-have but if the tool is aimed at beginners maybe it's not a real problem. Conversely, beginners could use the mega-yarn-file pattern and at the point where they have so many nodes that they need to start getting organized we could point them to a "using folders" document. I like workflows that "level up" as you get deeper into a project.

**Comment by pjarnfelt** [comment]
I think we are talking past each other. 
The node organization I suggested was to organized it just as you wish. Not related to scenes or characters. 
My organization in the image was based on a branching dialogue that I wanted contained within one foldable node. The Balloon Bush dialogue; where it branches to an A node and a B node (that then has 3 steps). This works for us when the dialogues are branching out, but of course wouldn't work when they are branching in. However, just being able to put all nodes pertaining to one conversation into a parent node would make organization much easier for us.
Having a folder for Bob and one for Sally makes no sense to me as it would be more logical for me to organize the yarn files based on levels in the game. 

- level 1
- level 2
- level 3

Level 1 would have
- Bob conversation 1
- - branching node A
- - branching node B
- Sally conversation 1
- Bob conversation when talked to sally
etc

**Comment by SimianLogic** [comment]
In my contrived example, Bob & Sally are different story branches (i.e. choose your character at the beginning of the game) and not characters within the same scene/level (which would look more like how you describe)--but I see what you mean. The macro-level project organization I'm suggesting doesn't solve the micro-level node hierarchy.

If you go the metadata route, it's going to be hard to not break things when you edit the yarn files manually. I wonder if you could solve this with a naming convention instead? Rewriting your scene above, your nodes in order would be something like:

1. L1_bob_convo
2. L1_bob_convo_branch_a
3. L1_bob_convo_branch_b
4. L1_sally_convo
5. L1_bob_after_sally

Since the tree view is built sequentially you could check to see if CURRENT_NODE's name was prefixed with PREVIOUS_NODE and treat that as a child/indentation.

- L1_bob_convo
- - branch_a
- - branch_b
- L1_sally_convo
- L1_bob_after_sally

If you dragged the *branch_a* node from *L1_bob_convo* to *L1_sally_convo* it would actually rename it from *L1_bob_convo_branch_a* to *L1_sally_convo_branch_a* (retaining its suffix and changing prefix). If you dragged it out of *L1_bob_convo* into the root hierarchy it would change from *L1_bob_convo_branch_a* to just *branch_a* (no prefix).

**Comment by radiatoryang** [comment]
so uh the node hierarchy stuff is just a side effect of hacking this from Unity's TreeModel example, which had parenting built into it... I think some of my concern with supporting it would be that it implies a relationship between nodes that isn't actually there? like, will people assume a parent node ending automatically passes control flow onto its children?

@SimianLogic : I think one problem with storing hierarchy as name prefixes / auto-renaming is that would break node links in the actual body text? or you think Merino should go through every node body and fix-up node name references?

to me, auto-renaming seems much more intrusive and text-editor-unfriendly than saving a Merino-only "ParentNode" tag to child nodes

but I think generally, I'm gravitating toward the need for both things: folder support, and also node hierarchy support

**Comment by SimianLogic** [comment]
yep i'm dumb. mucking with filenames is a bad idea. LMK if you want help on the folder support, I might have some time next week to tinker with that

**Comment by lordmortis** [comment]
If you wanted to do it, i'd suggest adding a Merino-specific tag to the header of the yarn node, something like this:
`parent: My Parent Node Name` 
?

**Comment by lordmortis** [comment]
(also if you do this, i'll be adding support to the Jacquard yarn parser. Not sure how we'll display it in the editor yet)
